pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,179 @@
+	userID string
+	logger log.Logger
+	reg    prometheus.Registerer
+
+	mtx    sync.Mutex
+	states map[string]cluster.State
+
+	replicationFactor  int
+	replicateStateFunc func(context.Context, string, *clusterpb.Part) error
+	positionFunc       func(string) int
+
+	partialStateMergesTotal  *prometheus.CounterVec
+	partialStateMergesFailed *prometheus.CounterVec
+	stateReplicationTotal    *prometheus.CounterVec
+	stateReplicationFailed   *prometheus.CounterVec
+
+	msgc   chan *clusterpb.Part
+	stopc  chan struct{}
+	readyc chan struct{}
+}
+
+// newReplicatedStates creates a new state struct, which manages state to be replicated between alertmanagers.
+func newReplicatedStates(userID string, rf int, f func(context.Context, string, *clusterpb.Part) error, pf func(string) int, stopc chan struct{}, l log.Logger, r prometheus.Registerer) *state {
+
+	s := &state{
+		logger:             l,
+		userID:             userID,
+		replicateStateFunc: f,
+		replicationFactor:  rf,
+		positionFunc:       pf,
+		states:             make(map[string]cluster.State, 2), // we use two, one for the notifications and one for silences.
+		msgc:               make(chan *clusterpb.Part),
+		readyc:             make(chan struct{}),
+		stopc:              stopc,
+		reg:                r,
+		partialStateMergesTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_total",
+			Help: "Number of times we have received a partial state to merge for a key.",
+		}, []string{"key"}),
+		partialStateMergesFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_failed_total",
+			Help: "Number of times we have failed to merge a partial state received for a key.",
+		}, []string{"key"}),
+		stateReplicationTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_total",
+			Help: "Number of times we have tried to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+		stateReplicationFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_failed_total",
+			Help: "Number of times we have failed to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+	}
+
+	go s.handleBroadcasts()
+	return s
+}
+
+// AddState adds a new state that will be replicated using the ReplicationFunc. It returns a channel to which the client can broadcast messages of the state to be sent.
+func (s *state) AddState(key string, cs cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return &stateChannel{
+		send: send,
+	}
```

Comment:
More of a question than anything, now I've spent more time walking through this code. (Apologies if I'm missing something obvious here...)

Do you think we'll need the flexibility of having the `stateChannel` being passed a function in this way? Could the `stateChannel` just take `s.msgc` and `key`, and the `Broadcast` function be implemented concretely? (Illustrated this in a separate comment)

```suggestion
	return &stateChannel{
		msgc: s.msgc,
		key: key,
	}
```